### PR TITLE
Control Plane, Data Plane, and Pausing

### PR DESCRIPTION
**Pull request by drinkcoffee**
* Revised text
* Added examples

Notes: 
* I have provided a real world example of issues when the proposed advice has not been taken. I think having real world examples will help people to understand the importance of a particular idea.
* For the examples, I linked to the analysis by rekt.news . I feel that this is a reliable source of detailed analysis that is likely to be around for some years. 
* When referring to "hacks", I have only called them "issues". I feel that this is less inflammatory.

### CODE REVIEW

File: docs/framework/20sources/30implementation/pause.md [hunk 1]
```diff
@@ -1,14 +1,45 @@
...
 attack.
 
-When analyzing whether a project can be paused, it is important 
-to check whether all data processing functions can be paused, or 
-just some parts of the project. The more of data processing 
-functions that can be paused, the less risky the project is.
+For Ethereum based projects, the OpenZeppelin project has an example 
+ contract [Pausable.sol](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/Pausable.sol)
+that can be used to implement pausing. Using this template, 
+pausing a function becomes as simple as adding a modifier **whenNotPaused**.
```
Comment (ermyas) [inline docs/framework/20sources/30implementation/pause.md:14]: nit: suggest we use in-line code markers (i.e. backticks ` ) when referring to functions as opposed to bold/italics.

File: docs/framework/20sources/30implementation/pause.md [hunk 2]
```diff
@@ -1,14 +1,45 @@
...
+    }
+    function transfer(
+        address _sender,
+        address _tokenContract,
+        address _recipient,
+        uint256 _amount
+    ) external whenNotPaused {
+        // Only executed when not paused
+```
+
```
Comment (ermyas) [inline docs/framework/20sources/30implementation/pause.md:30]: nit: not sure providing this example is necessary considering the target audience for this framework.

File: docs/framework/20sources/30implementation/pause.md [hunk 3]
```diff
@@ -1,14 +1,45 @@
...
+When analyzing whether a project can be paused, it is important
+to check whether all data processing functions can be paused, or
```
Comment (ermyas) [inline docs/framework/20sources/30implementation/pause.md:32]: nit: "all data processing functions" -> maybe we should qualify this as all _public_ or _external_ data processing functions?

File: docs/framework/20sources/30implementation/pause.md [hunk 4]
```diff
@@ -1,14 +1,45 @@
...
+just some parts of the project. 
+
+
+***
+#### Example
+For example, in [August 2022 the Nomad Bridge](https://rekt.news/nomad-rekt/) had an issue.
```
Comment (ermyas) [inline docs/framework/20sources/30implementation/pause.md:38]: I think the rekt summary of this incident contains subtle inaccuracies. I would suggest we cite the official [RCA from the Nomad team](https://medium.com/nomad-xyz-blog/nomad-bridge-hack-root-cause-analysis-875ad2e5aacd) or Coinbase's incident analysis which offers a more accurate explanation of the vulnerability.

File: docs/framework/20sources/30implementation/mixing-control-data-flow.md [hunk 5]
```diff
@@ -1,15 +1,46 @@
-### Mixing of Control and Data Flow
+### Mixing of Control and Data Plane
+
+The terms *Control Plane* and *Data Plane* come from networking [[1](https://en.wikipedia.org/wiki/Control_plane)]. In the
```
Comment (ermyas) [inline docs/framework/20sources/30implementation/mixing-control-data-flow.md:3]: nit: the reference number is static and likely to get out of sync, suggest removing and associating the link to a relevant phrase.

File: docs/framework/20sources/30implementation/pause.md [hunk 6]
```diff
@@ -1,14 +1,45 @@
 ### Ability to Pause Project
```
Comment (ermyas) [inline docs/framework/20sources/30implementation/pause.md:1]: Is this more of an operational concern than an implementation one? should this be moved? I can see arguments both ways, but it feels more like a source of operational risk than an implementation one to me.

**Review by ermyas — APPROVED** [review]

File: docs/framework/20sources/30implementation/pause.md (hunk 1 above)
Comment (drinkcoffee) [reply docs/framework/20sources/30implementation/pause.md]: done

File: docs/framework/20sources/30implementation/pause.md (hunk 2 above)
Comment (drinkcoffee) [reply docs/framework/20sources/30implementation/pause.md]: I think we will have quite a varied audience for this work. As such, having relevant examples will be useful.

File: docs/framework/20sources/30implementation/pause.md (hunk 3 above)
Comment (drinkcoffee) [reply docs/framework/20sources/30implementation/pause.md]: It applies to all functions. A private or internal function is only going to be accessible via an external or public function. The pausing capability can happen at any point in the logic of the code - not necessarily at the entry point.

File: docs/framework/20sources/30implementation/pause.md (hunk 4 above)
Comment (drinkcoffee) [reply docs/framework/20sources/30implementation/pause.md]: Ermyas and I had a discussion about this offline. Having the official source of the root cause analysis / post mortem should be the primary source. However, having other secondary sources, such as rekt can be helpful, as some times teams' internal analyses gloss over some issues.

File: docs/framework/20sources/30implementation/mixing-control-data-flow.md (hunk 5 above)
Comment (drinkcoffee) [reply docs/framework/20sources/30implementation/mixing-control-data-flow.md]: Good point. I replaced with the source - Wikipedia.

File: docs/framework/20sources/30implementation/pause.md (hunk 6 above)
Comment (drinkcoffee) [reply docs/framework/20sources/30implementation/pause.md]: There is information about pausing (from a design perspective) in implementation, and then separately a section in operation, discussing the operational concerns.

**Pull request merged**